core/iwasm/libraries/lib-pthread/lib_pthread_wrapper.c

```diff
@@ -1085,6 +1122,169 @@ posix_memalign_wrapper(wasm_exec_env_t exec_env, void **memptr, int32 align,
     return 0;
 }
 
+#ifdef WASM_ENABLE_LIB_PTHREAD_SEMAPHORE
+
+static int32
+sem_open_wrapper(wasm_exec_env_t exec_env, const char *name, int32 oflags)
+{
+    korp_sem *psem = NULL;
+    ThreadInfoNode *info_node = NULL;
+
+    /**
+     * For RTOS, global semaphore map is safety for share the same semaphore
+     * between task/pthread.
+     * For Unix like system, it's dedicate for multiple processes.
+     */
+
+    if ((info_node = bh_hash_map_find(sem_info_map, (void *)name))) {
+        return info_node->handle;
+    }
+
+    if (!(psem = os_sem_open(name, oflags))) {
+        goto fail1;
+    }
+
+    if (!(info_node = wasm_runtime_malloc(sizeof(ThreadInfoNode))))
+        goto fail2;
+
+    memset(info_node, 0, sizeof(ThreadInfoNode));
+    info_node->exec_env = exec_env;
+    info_node->handle = allocate_handle();
+    info_node->type = T_SEM;
+    info_node->u.sem = psem;
+    info_node->status = SEM_CREATED;
+
+    if (!append_thread_info_node(info_node))
+        goto fail3;
+
+    if (!bh_hash_map_insert(sem_info_map, (void *)name, info_node)) {
```

Comment:
An issue for multi-thread: suppose thread A opens a semaphore with a name, and the info_node is added into both thread_info_map and sem_info_map, and B opens a semaphore with the same name, this time it will just return success in L1140 as is has been created. And then thread A exits, the semaphore may be destroyed during thread exiting, but thread B still uses it, which might cause core dump.

Reply:
In this case, thread B take a invalid handle (fd number), and call sem_xxx will get a return value indicate the error

Reply:
OK